manager/controllers/app/blueprint_controller.go

```diff
@@ -163,26 +167,53 @@ func (r *BlueprintReconciler) reconcile(ctx context.Context, log logr.Logger, bl
 		// Add metadata arguments (namespace, name, labels, etc.)
 		args["metadata"] = map[string]interface{}{
 			"name":      utils.Hash(step.Name, 20),
-			"namespace": blueprint.Namespace, // TODO(roee.shlomo): should eventually be a dedicated namespace per M4DApplication
+			"namespace": blueprint.Namespace,
 			"labels":    blueprint.Labels,
 		}
+		// we add the "r" character at the beginning of the release name, since it must begin with an alphabetic character
+		releaseName := "r" + args["metadata"].(map[string]interface{})["name"].(string)
+		log.V(0).Info("Release name: " + releaseName)
+		// check if the blueprint is about to be deleted
+		if blueprint.DeletionTimestamp != nil {
+			_, _ = r.deleteChartResource(ctx, blueprint.Namespace, releaseName)
+			continue
+		}
 
-		// Process templates with arguments
-		for _, resource := range templateSpec.Resources {
-			if blueprint.DeletionTimestamp != nil {
-				_, _ = r.deleteChartResource(ctx, log, resource, args)
-				continue
+		numReleases++
+		//check the release status
+		rel, err := r.Helmer.Status(blueprint.Namespace, releaseName)
+		// unexisting release or a failed release - re-apply the chart
+		if updateRequired || err != nil || rel == nil || rel.Info.Status == release.StatusFailed {
+			// Process templates with arguments
+			for _, resource := range templateSpec.Resources {
+				if _, err := r.applyChartResource(ctx, log, resource, args, blueprint.Namespace, releaseName); err != nil {
+					blueprint.Status.Error += errors.Wrap(err, "ChartDeploymentFailure: ").Error() + "\n"
+				}
 			}
-
-			if res, err := r.applyChartResource(ctx, log, resource, args); err != nil {
-				r.Log.V(0).Info("failed to apply chart: " + err.Error())
-				return res, err
+		} else if rel.Info.Status == release.StatusDeployed {
+			// TODO: add release notes of the read module to the status
+			if args["flow"] == "read" {
+				log.V(0).Info(rel.Info.Notes)
+			}
+			status, errMsg := r.checkReleaseStatus(releaseName, blueprint.Namespace)
+			if status == corev1.ConditionFalse {
+				blueprint.Status.Error += "ResourceAllocationFailure: " + errMsg + "\n"
+			} else if status == corev1.ConditionTrue {
+				numReady++
 			}
 		}
 	}
+	// check if all releases reached the ready state
+	if numReady == numReleases {
+		// all modules have been orhestrated successfully - the data is ready for use
+		blueprint.Status.Ready = true
+		return ctrl.Result{}, nil
+	}
 
-	log.V(0).Info("PILOT_MAGIC_NUMBER=" + os.Getenv("PILOT_MAGIC_NUMBER"))
-	blueprint.Status.Ready = false // Won't be ready until all components are running.  TODO - add status checks
+	// the status is unknown yet - continue polling
+	if blueprint.Status.Error == "" {
```

Comment:
Is that the error message field? Because if so then it's optional meaning this won't work.

Reply:
This is the blueprint status field, not the resource status indicator. In this function it acts as a local variable rather than the real status field, as it is initialized and calculated here, while it is used to update the blueprint status later in Reconcile() method.